### PR TITLE
Add options to add in UBOOT

### PR DESCRIPTION
**Pull request by ruchi393**

### CODE REVIEW

**Comment by ruchi393** [comment]
This PR is dependent on 
https://github.com/OP-TEE/manifest/pull/185

File: qemu_v8.mk [hunk 1]
```diff
@@ -41,15 +47,62 @@ QEMU_PATH		?= $(ROOT)/qemu
...
+BL33_BIN		?= $(EDK2_BIN)
+BL33_DEPS		?= edk2
+endif
 
 ################################################################################
 # Targets
 ################################################################################
-all: arm-tf buildroot edk2 linux optee-os qemu soc-term
-clean: arm-tf-clean buildroot-clean edk2-clean linux-clean optee-os-clean \
+TARGET_DEPS ?= arm-tf buildroot linux optee-os qemu soc-term
```
Comment (jenswi-linaro) [inline qemu_v8.mk:86]: Does it really make sense to have this overridable?

Overall I think that most of the `?=` locations in these make files should have `:=` or `=` instead. There are only few variables which it makes sense to override. Fixing this is out of scope for this PR, but there's no need to make matters worse.

File: qemu_v8.mk [hunk 2]
```diff
@@ -141,6 +195,37 @@ edk2: edk2-common
 
 edk2-clean: edk2-clean-common
 
+################################################################################
```
Comment (jenswi-linaro) [inline qemu_v8.mk:198]: This is perhaps something that could be moved to `common.mk` instead. There's quite a few platforms now compiling U-Boot using a somewhat similar pattern, probably due to copy paste + some small tweak in each case.

File: qemu_v8.mk [hunk 3]
```diff
@@ -189,6 +274,35 @@ soc-term:
 soc-term-clean:
 	$(MAKE) -C $(SOC_TERM_PATH) clean
 
+################################################################################
+# mkimage - create images to be loaded by U-boot
+################################################################################
+# Without the objcopy, the uImage will be 10x bigger.
+uimage: linux $(BINARIES_PATH)
```
Comment (jenswi-linaro) [inline qemu_v8.mk:281]: Why not make the target `$(KERNEL_UIMAGE)`? It seems that `uimage` will always be out of date.

File: qemu_v8.mk [hunk 4]
```diff
@@ -189,6 +274,35 @@ soc-term:
...
+	${AARCH64_CROSS_COMPILE}objcopy -O binary \
+					-R .note \
+					-R .comment \
+					-S $(LINUX_PATH)/vmlinux \
+					$(BINARIES_PATH)/linux.bin && \
```
Comment (jenswi-linaro) [inline qemu_v8.mk:286]: No need for `&& \` at the end.

File: qemu_v8.mk [hunk 5]
```diff
@@ -141,6 +195,37 @@ edk2: edk2-common
...
+
+.PHONY: uboot-menuconfig
+uboot-menuconfig: uboot-defconfig
+	$(MAKE) -C $(UBOOT_PATH) \
+		CROSS_COMPILE="$(CCACHE)$(AARCH64_CROSS_COMPILE)" \
+		menuconfig
+
+.PHONY: uboot-clean
+uboot-clean:
+	cd $(UBOOT_PATH) && git clean -xdf
```
Comment (jenswi-linaro) [inline qemu_v8.mk:223]: U-Boot already have a `clean` target.

File: qemu_v8.mk [hunk 6]
```diff
@@ -189,6 +274,35 @@ soc-term:
...
+	$(MKIMAGE_PATH)/mkimage -A arm64 \
+				-O linux \
+				-T kernel \
+				-C none \
+				-a $(KERNEL_LOADADDR) \
+				-e $(KERNEL_ENTRY) \
+				-n "Linux kernel" \
+				-d $(BINARIES_PATH)/linux.bin $(KERNEL_UIMAGE)
+
+urootfs: buildroot $(BINARIES_PATH)
```
Comment (jenswi-linaro) [inline qemu_v8.mk:296]: Shouldn't the target be ` $(ROOTFS_UGZ)`?

File: qemu_v8.mk [hunk 7]
```diff
@@ -189,6 +274,35 @@ soc-term:
 soc-term-clean:
 	$(MAKE) -C $(SOC_TERM_PATH) clean
 
+################################################################################
+# mkimage - create images to be loaded by U-boot
+################################################################################
+# Without the objcopy, the uImage will be 10x bigger.
```
Comment (jenswi-linaro) [inline qemu_v8.mk:280]: Aha, good to know!

File: qemu_v8.mk (hunk 1 above)
Comment (ruchi393) [reply qemu_v8.mk:86]: I will change TARGET_DEPS ?= to TARGET_DEPS :=

File: qemu_v8.mk (hunk 2 above)
Comment (ruchi393) [reply qemu_v8.mk:198]: Ok, will do

File: qemu_v8.mk (hunk 4 above)
Comment (ruchi393) [reply qemu_v8.mk:286]: I will remove that

**Comment by jforissier** [comment]
Cosmetic issues...
- It would be nice to state how to enable U-Boot in the commit description of "qemu_v8.mk: Add option to use uboot instead of edk2". Adding "(make UBOOT=y)" at the end of the sentence would be good enough I think.
- U-Boot is spelt U-Boot, not uboot or UBOOT.

**Comment by jforissier** [comment]
Is GICv3 needed for U-Boot support? I don't think so, therefore commit "qemu_v8.mk: Add option to compile with GICv3" probably shouldn't be in this PR right?

**Comment by ruchi393** [comment]
> Is GICv3 needed for U-Boot support? I don't think so, therefore commit "qemu_v8.mk: Add option to compile with GICv3" probably shouldn't be in this PR right?

Yes, you are right. I will split the PR up.

**Comment by ruchi393** [comment]
Fixed based on comments on U-Boot patch. 
Removed GICv3 patch from the PR.

File: common.mk [hunk 8]
```diff
@@ -361,6 +362,26 @@ edk2-clean-common:
...
+uboot-defconfig: $(UBOOT_PATH)/.config
+
+.PHONY: uboot-common
+uboot-common: uboot-defconfig
+	$(MAKE) -C $(UBOOT_PATH) $(UBOOT_COMMON_FLAGS)
+
+.PHONY: uboot-clean-common
+uboot-clean-common:
+	$(MAKE) -C $(UBOOT_PATH) $(UBOOT_COMMON_FLAGS) distclean
+
```
Comment (jforissier) [inline common.mk:384]: 1. I think all platforms should use these targets as soon as they are introduced. How bout one refactoring commit with the changes to `common.mk` and all platform files ("Move u-boot* targets to common.mk"), then one separate commit for the new QEMUv8 stuff?
2. Why `uboot*` when `qemu.mk` already uses the spelling `u-boot*`?

File: qemu_v8.mk [hunk 9]
```diff
@@ -23,6 +23,9 @@ include common.mk
 
 DEBUG ?= 1
 
+# Option to use uboot in the boot flow instead of EDK2
```
[the remark this answers is not in the document]
Comment (ruchi393) [reply qemu_v8.mk:26]: Sorry, I missed to replace this one.

File: kconfigs/u-boot_qemu_v8.conf [hunk 10]
```diff
@@ -0,0 +1,4 @@
+CONFIG_SYS_TEXT_BASE=0x60000000
+CONFIG_BOOTCOMMAND="smhload uImage ${kernel_addr_r} && smhload rootfs.cpio.uboot ${ramdisk_addr_r} &&  setenv bootargs console=ttyAMA0,115200 earlyprintk=serial,ttyAMA0,115200 root=/dev/ram && bootm ${kernel_addr_r} ${ramdisk_addr_r} ${fdt_addr}"
+CONFIG_SEMIHOSTING=y
+
```
[the remark this answers is not in the document]
Comment (ruchi393) [reply kconfigs/u-boot_qemu_v8.conf:4]: Will remove

File: qemu_v8.mk [hunk 11]
```diff
@@ -189,6 +243,35 @@ soc-term:
 soc-term-clean:
 	$(MAKE) -C $(SOC_TERM_PATH) clean
 
+################################################################################
+# mkimage - create images to be loaded by U-Boot
+################################################################################
+# Without the objcopy, the uImage will be 10x bigger.
+$(KERNEL_UIMAGE): linux $(BINARIES_PATH)
```
[the remark this answers is not in the document]
Comment (ruchi393) [reply qemu_v8.mk:250]: Ok

File: qemu_v8.mk [hunk 12]
```diff
@@ -189,6 +243,35 @@ soc-term:
...
+	$(MKIMAGE_PATH)/mkimage -A arm64 \
+				-O linux \
+				-T kernel \
+				-C none \
+				-a $(KERNEL_LOADADDR) \
+				-e $(KERNEL_ENTRY) \
+				-n "Linux kernel" \
+				-d $(BINARIES_PATH)/linux.bin $(KERNEL_UIMAGE)
+
+$(ROOTFS_UGZ): buildroot $(BINARIES_PATH)
```
[the remark this answers is not in the document]
Comment (ruchi393) [reply qemu_v8.mk:265]: Ok

File: common.mk (hunk 8 above)
Comment (ruchi393) [reply common.mk:384]: > 1. I think all platforms should use these targets as soon as they are introduced. How about one refactoring commit with the changes to `common.mk` and all platform files ("Move u-boot* targets to common.mk"), then one separate commit for the new QEMUv8 stuff?

Can I just add u-boot target in QEMUv8 file for now as I did in earlier PR. I need this added to get the xen workflow added. Then I can work on a separate PR to move things to common.mk and fix all platforms. Would that work ?

> 2. Why `uboot*` when `qemu.mk` already uses the spelling `u-boot*`?

 No particular reason. I actually didn't see the qemu.mk. I can change to u-boot for consistency.

Comment (jforissier) [reply common.mk:384]: > Then I can work on a separate PR to move things to common.mk and fix all platforms. Would that work ?

Yes, totally acceptable.

> No particular reason. I actually didn't see the qemu.mk. I can change to u-boot for consistency.

Yes, please. No big deal, TBH I would have chosen `uboot` too, but consistency is more important.

Comment (jbech-linaro) [reply common.mk:384]: > > No particular reason. I actually didn't see the qemu.mk. I can change to u-boot for consistency.
> 
> Yes, please. No big deal, TBH I would have chosen `uboot` too, but consistency is more important

I could be the one to blame for that. I think I've tried to use the correct spelling. But if it was today, I probably would've went with `uboot` as well. (and I'm glad we use `optee_xyz` for gits etc and not `op-tee_xyx`).

**Comment by ruchi393** [comment]
Added a fix commit to address all comments. Please help review.

File: qemu_v8.mk [hunk 13]
```diff
@@ -41,15 +44,54 @@ QEMU_PATH		?= $(ROOT)/qemu
...
+TARGET_DEPS 		+= $(BL33_DEPS)
+
+ifeq ($(UBOOT),y)
+TARGET_DEPS		+= $(KERNEL_UIMAGE) $(ROOTFS_UGZ)
+TARGET_CLEAN		+= u-boot-clean
+else
+TARGET_CLEAN		+= edk2-clean
+endif
+
+all: $(TARGET_DEPS)
```
Comment (jenswi-linaro) [inline qemu_v8.mk:88]: These dependencies without actions can be repeated as many times as needed so there's generally not a need to do this trick with a variable. There's not much harm in doing like this either so keep this if you prefer that.

File: qemu_v8.mk [hunk 14]
```diff
@@ -189,6 +256,35 @@ soc-term:
 soc-term-clean:
 	$(MAKE) -C $(SOC_TERM_PATH) clean
 
+################################################################################
+# mkimage - create images to be loaded by U-Boot
+################################################################################
+# Without the objcopy, the uImage will be 10x bigger.
+$(KERNEL_UIMAGE): linux | $(BINARIES_PATH)
```
Comment (jforissier) [inline qemu_v8.mk:263]: `$(KERNEL_UIMAGE): u-boot linux | $(BINARIES_PATH)`

(`mkimage` is generated by target `u-boot`)

File: qemu_v8.mk [hunk 15]
```diff
@@ -189,6 +256,35 @@ soc-term:
...
+	$(MKIMAGE_PATH)/mkimage -A arm64 \
+				-O linux \
+				-T kernel \
+				-C none \
+				-a $(KERNEL_LOADADDR) \
+				-e $(KERNEL_ENTRY) \
+				-n "Linux kernel" \
+				-d $(BINARIES_PATH)/linux.bin $(KERNEL_UIMAGE)
+
+$(ROOTFS_UGZ): buildroot | $(BINARIES_PATH)
```
Comment (jforissier) [inline qemu_v8.mk:278]: `$(ROOTFS_UGZ): u-boot buildroot | $(BINARIES_PATH)`

(`mkimage` is generated by target `u-boot`)

**Review by jforissier — COMMENTED** [review]
One issue I found when building with `make -j10`. With that fixed:

`Reviewed-by: Jerome Forissier <jerome@forissier.org>`
`Tested-by: Jerome Forissier <jerome@forissier.org>`

File: qemu_v8.mk (hunk 15 above)
Comment (ruchi393) [reply qemu_v8.mk:278]: Thanks @jforissier , I will add the dependency

File: qemu_v8.mk (hunk 13 above)
Comment (ruchi393) [reply qemu_v8.mk:88]: I just found it a little more readable hence did it this way.

**Comment by ruchi393** [comment]
Patches squashed and tag applied

**Comment by ruchi393** [comment]
@jforissier , @jenswi-linaro , thanks for the review. All comments have been addressed. Please have a look.

**Review by jforissier — APPROVED** [review]
LGTM

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by ruchi393** [comment]
Tags applied

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by ruchi393** [comment]
Tags applied

**Pull request merged**